File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxSum = INT_MIN;

    int dfs(TreeNode* node)
    {
        if (!node) return 0;

        int leftMax = max(dfs(node->left), 0);

        int rightMax = max(dfs(node->right), 0);

        maxSum = max(maxSum, node->val + leftMax + rightMax);

        return node->val + max(leftMax, rightMax);
    }
    int maxPathSum(TreeNode* root) {
        dfs(root);
        return maxSum;
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        
        res = [root.val]
        
        # function to calc max sum without split
        def dfs(root):
            if not root:
                return 0
            
            leftMax = dfs(root.left)
            rightMax = dfs(root.right)
            
            # Avoiding negative vals
            leftMax = max(leftMax, 0)
            rightMax = max(rightMax, 0)
            
            # computing the result with split
            res[0] = max(res[0], root.val + leftMax + rightMax)
            
            
            # returning the max sum without split
            
            return root.val + max(leftMax, rightMax)
        
        dfs(root)
        return res[0]